src/librustc_typeck/check/coercion.rs

```diff
@@ -355,6 +355,20 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
 
         // Now apply the autoref. We have to extract the region out of
         // the final ref type we got.
+        if ty == a && mt_a.mutbl == hir::MutImmutable && autoderefs == 1 {
```

Comment:
> It's a bit subtle, but I think this is valid only because mt_a.mutbl is the source mutability and from immutable you can only go to immutable.

I guess that was the idea, yeah. I can make the comment more explicit and add an assert. :)